upx-recovery-tool.py

```diff
@@ -0,0 +1,289 @@
+        self.in_fd = open(self.in_file, "rb")
+        self.elf = ELFFile(self.in_fd)
+
+        # Check if is UPX
+        if not self.is_upx():
+            raise NonUpxError
+
+        # Get UPX version. p_info fix doesn't work with UPX 4
+        self.detect_version()
+
+        # All checks passed
+        shutil.copy(in_file, out_file)
+
+        self.out_fd = open(out_file, "r+b")
+        self.buff = mmap.mmap(self.out_fd.fileno(), 0)
+
+        self.load_structs()
+
+    def check_file_type(self):
+        """ Method to check if the class will be able to analyze this type of file """
+
+        if not os.path.isfile(self.in_file):
+            raise UnsupportedFileError("Is not a file")
+
+        # Check magic filetype
+        magic_str = magic.from_file(self.in_file)
+        
+        if magic_str.startswith("ELF "):
+            self.exe_type = "ELF"
+            magic_arch = magic_str.split(',')[1].strip()
+
+            if magic_arch in self.upx_sigs.keys():
+                self.arch = magic_arch
+                return
+
+        raise UnsupportedFileError(f"Unsupported file type '{magic_str}'")
+
+    def is_upx(self):
+        """ Method that looks for ASM signatures to identify an UPX executable """
+
+        # Instead of looking for the pattern in all the executable bytes, look for the
+        # UPX sigs in the EP. The code is not so beauty, but it'll be more efficient.
+        ep_bytes = self.get_ep_bytes(50)
+
+        for sig in self.upx_sigs[self.arch]:
+            if re.match(sig, ep_bytes, re.DOTALL):
+                return True
+
+        return False
+
+    def detect_version(self):
+        """ Method to know the UPX version used to pack the executable """
+        # TODO: The current detection method is very naive and can be easily tricked
+
+        # Load file in memory to look for string with version
+        self.in_fd.seek(0)
+        in_buff = mmap.mmap(self.in_fd.fileno(), 0, access=mmap.ACCESS_READ)
+
+        # Detect UPX version
+        #if in_buff.find(b"$Id: UPX 4.00") != -1:
+        version_off = in_buff.find(b"$Id: UPX ")
+        if version_off == -1:
+            print("[!] UPX version could not be detected")
+            self.version = None
+        else:
+            self.version = int(in_buff[version_off + 9: version_off + 10])
+            # TODO: Use 're' to get major and minor version
+
+        in_buff.close()
+
+    def load_structs(self):
+
+        eh = self.elf.header
+
+        # l_info
+        self.l_info_off = eh.e_phoff + eh.e_phnum * eh.e_phentsize
+        self.l_info = l_info_s(self.buff[self.l_info_off: self.l_info_off + 12])
+
+        # p_info
+        self.p_info_off = self.l_info_off + 12
+        self.p_info = p_info_s(self.buff[self.p_info_off: self.p_info_off + 12])
+
+    def patch(self, patch_bytes, offset):
+        """ Method to patch bytes in the output binary """
+
+        self.buff[offset: offset + len(patch_bytes)] = patch_bytes
```

Comment:
I would add a check that the offset + len are not bigger than the total buff's size